changelog.d/13591.misc

```diff
@@ -0,0 +1 @@
+Improve performance of `@cachedList`.
```

Comment:
Great progress @erikjohnston 🎉

Here are the before/after timings for this PR using the `have_seen_events` benchmark from https://github.com/matrix-org/synapse/pull/13561

\# events | Before | After
--- | --- | ---
50k | . <pre>Benchmark time (1 cold cache ): 3.7170820236206055<br>Benchmark time (2, warm cache): 0.2985079288482666<br>Benchmark time (3, warm cache): 0.28847789764404297<br>Benchmark time (4, odds      ): 0.1537461280822754<br>Benchmark time (5, odds      ): 0.14780497550964355<br>Benchmark time (6, evens     ): 0.1475691795349121<br>Benchmark time (7, evens     ): 0.14868617057800293</pre> | . <pre>Benchmark time (1 cold cache ): 1.062384843826294<br>Benchmark time (2, warm cache): 0.18310809135437012<br>Benchmark time (3, warm cache): 0.17781400680541992<br>Benchmark time (4, odds      ): 0.09939789772033691<br>Benchmark time (5, odds      ): 0.09195113182067871<br>Benchmark time (6, evens     ): 0.08882665634155273<br>Benchmark time (7, evens     ): 0.08661413192749023</pre>
100k | . <pre>Benchmark time (1 cold cache ): 8.10055136680603<br>Benchmark time (2, warm cache): 0.6121761798858643<br>Benchmark time (3, warm cache): 0.6093218326568604<br>Benchmark time (4, odds      ): 0.29950785636901855<br>Benchmark time (5, odds      ): 0.3049640655517578<br>Benchmark time (6, evens     ): 0.3025388717651367<br>Benchmark time (7, evens     ): 0.29833483695983887</pre> | . <pre>Benchmark time (1 cold cache ): 2.264657735824585<br>Benchmark time (2, warm cache): 0.3539161682128906<br>Benchmark time (3, warm cache): 0.3556997776031494<br>Benchmark time (4, odds      ): 0.17580008506774902<br>Benchmark time (5, odds      ): 0.17888712882995605<br>Benchmark time (6, evens     ): 0.1756150722503662<br>Benchmark time (7, evens     ): 0.17418384552001953</pre>
200k | . <pre>Benchmark time (1 cold cache ): 19.106724977493286<br>Benchmark time (2, warm cache): 22.98161005973816<br>Benchmark time (3, warm cache): 23.126408100128174<br>Benchmark time (4, odds      ): 11.401129007339478<br>Benchmark time (5, odds      ): 0.6159579753875732<br>Benchmark time (6, evens     ): 12.087002992630005<br>Benchmark time (7, evens     ): 0.6241748332977295</pre> | . <pre>Benchmark time (1 cold cache ): 4.827088117599487<br>Benchmark time (2, warm cache): 4.958348035812378<br>Benchmark time (3, warm cache): 4.923892974853516<br>Benchmark time (4, odds      ): 2.5139119625091553<br>Benchmark time (5, odds      ): 0.3530576229095459<br>Benchmark time (6, evens     ): 2.465486764907837<br>Benchmark time (7, evens     ): 0.3511021137237549</pre>